src/Lucene.Net/Search/FieldCacheImpl.cs

```diff
@@ -259,9 +231,25 @@ public virtual System.Object Get(IndexReader reader, Entry key, IState state)
                         return progress.value_Renamed;
                     }
                 }
-                return value;
+
+                return (T) value;
             }
-            
+
+            public void AddToBag<TType>(ConcurrentBag<CacheEntry> concurrentBag)
+            {
+                foreach (var readerCacheEntry in readerCache)
+                {
+                    var readerKey = readerCacheEntry.Key;
+                    var innerCache = readerCacheEntry.Value;
+                    foreach (var mapEntry in innerCache)
+                    {
+                        Entry entry = mapEntry.Key;
```

Comment:
Need to check that this isn't null, no?

Reply:
This is inner key, so it can't be null. But I think we do need to check the `readerKey` for `null`

Reply:
`innerCache` is also `WeakTable`, no?

Reply:
It was a `HashMap`, but changed it to `ConcurrentDictionary` to reduce some locks